File: src/scenes/board_scene/pattern_card.js
import Phaser from "phaser";
import { scale_marble_size } from "./board_utils.bs";
const GRID_W = 3;
const GRID_H = 3;

export default class PatternCard {
  constructor({ pattern, scene, x = 0, y = 0, width = 0, heigth = 0 }){
    this.isDone = false;
    this.pattern = pattern;
    this.scene = scene;
    this.drawCard = this.drawCard.bind(this);
    this.x = x;
    this.y = y;
    this.width = width;
    this.heigth = heigth;
  }
  
  drawDoneOverlay(){
    if(!this.isDone){
      this.isDone = true;
      const overlay = this.scene.add.graphics();
      const doneSprite = this.scene.add.sprite(this.width * 0.5, this.heigth * 0.5, "check");
      doneSprite.setOrigin(0.5, 0.5);
      doneSprite.displayHeight = this.heigth * 0.3;
      doneSprite.displayWidth = this.width * 0.3;
      overlay.fillStyle(parseInt("757575", 16), 0.5);
      overlay.fillRect(0,0,this.width,this.heigth);
      this.cardContainer.add(overlay)
      this.cardContainer.add(doneSprite);

    }
  }
  drawCard(){
    this.cardContainer = this.scene.add.container();
    const sWidth = this.width / GRID_W;
    const sHeight = this.heigth / GRID_H;
    const [mWidth, mHeight] = scale_marble_size(sWidth, sHeight, 30);

    for (let i = 0; i < GRID_H; i++) {
      for (let j = 0; j < GRID_W; j++) {
        const slot = this.scene.add.graphics();
        const sX = j * sWidth;
        const sY = i * sHeight;
        slot.lineStyle(5, parseInt("E1E2E1", 16), 1.0);
        slot.fillStyle(parseInt("bbdefb", 16), 1.0);
        slot.beginPath();
        //draw top border if needed
        if (i === 0) {
          slot.moveTo(sX, sY);
          slot.lineTo(sX + sWidth, sY);
        }
        //draw right and bottom border
        slot.moveTo(sX + sWidth, sY);
        slot.lineTo(sX + sWidth, sY + sHeight);
        slot.lineTo(sX, sY + sHeight);
        //draw left border if needed
        if (j === 0) {
          slot.lineTo(sX, sY);
        }
        slot.fillRect(j * sWidth, i * sHeight, sWidth, sHeight);
        slot.strokePath();
        this.cardContainer.add(slot);
        if(this.pattern[i][j] === 1){
          const marbleSprite = this.scene.add.sprite(sX + sWidth * 0.5, sY + sHeight * 0.5, "marble");
          this.cardContainer.add(marbleSprite);
          marbleSprite.displayHeight = mHeight;
          marbleSprite.displayWidth = mWidth;
        }
      }
    }
    this.cardContainer.setPosition(this.x, this.y);
  }
}

File: src/ml/pattern_card.ml
open Phaser
open GameObject
open Utils 

let grid_h = 3 

let grid_w = 3

let create_marble factory x y w h = 
  let marble = sprite factory x y "marble" in
  Sprite.displayWidthSet marble w;
  Sprite.displayHeightSet marble h;
  marble

let draw_card scene pattern x y w h = 
  Js.log2 "pattern" pattern;

  let object_factor = Scene.addGet scene in
  let card_container = container object_factor in
  let grid_w_f = float_of_int grid_w in
  let grid_h_f = float_of_int grid_h in
  let s_width = w /. grid_w_f in
  let s_height = h /. grid_h_f in
  let {width = m_width; height = m_height} = Utils.scale_marble_size s_width s_height 30.0 in
  for i = 0 to int_of_float grid_w_f - 1 do
    for j = 0 to int_of_float grid_h_f - 1 do
      let slot = graphics object_factor in
      let apply f = slot |. f in
      let s_x = float_of_int j *. s_width in
      let s_y = float_of_int i *. s_height in
      apply Graphics.lineStyle 5.0 (parse_int "E1E2E1" 16) 1.0 ;
      apply Graphics.fillStyle (parse_int "bbdefb" 16) 1.0 ;
      apply Graphics.beginPath ;
      (* draw top border if needed *)
      if i = 0 then (
        apply Graphics.moveTo s_x s_y ;
        apply Graphics.lineTo (s_x +. s_width) s_y ) ;
      (* draw right and bottom border *)
      apply Graphics.moveTo (s_x +. s_width) s_y ;
      apply Graphics.lineTo (s_x +. s_width) (s_y +. s_height) ;
      apply Graphics.lineTo s_x (s_y +. s_height) ;
      if j = 0 then apply Graphics.lineTo s_x s_y ;
      apply Graphics.fillRect
        (float_of_int j *. s_width)
        (float_of_int i *. s_height)
        s_width s_height ;
      apply Graphics.strokePath ;
      add_container card_container (`Graphics slot) ;
      Js.log3 i j pattern;
      if get_unsafe pattern i j = 1.0 then (
        let marble_sprite = create_marble object_factor 
          (s_x +. s_width *. 0.5) (s_y +. s_height *. 0.5) m_width m_height in 
        add_container card_container (`Sprite marble_sprite)  
        )
    done
  done ;
  Container.set_position card_container x y